languages/go/types/polar_types.go

```diff
@@ -15,6 +15,593 @@ type Call struct {
+}
+type ConstraintKindEq struct{}
+
+func (ConstraintKindEq) isConstraintKind() {}
+
+type ConstraintKindIn struct{}
+
+func (ConstraintKindIn) isConstraintKind() {}
+
+type ConstraintKindContains struct{}
+
+func (ConstraintKindContains) isConstraintKind() {}
+
+type ConstraintKindNeq struct{}
+
+func (ConstraintKindNeq) isConstraintKind() {}
+
+type ConstraintKindNin struct{}
+
+func (ConstraintKindNin) isConstraintKind() {}
+
+// ConstraintKind enum
+type ConstraintKindVariant interface {
+	isConstraintKind()
+}
+
+type ConstraintKind struct {
+	ConstraintKindVariant
+}
+
+func (result *ConstraintKind) UnmarshalJSON(b []byte) error {
+	var variantName string
+	var variantValue *json.RawMessage
+
+	// try and deserialize as a string first
+	err := json.Unmarshal(b, &variantName)
+	if err != nil {
+		var rawMap map[string]json.RawMessage
+		err := json.Unmarshal(b, &rawMap)
+		if err != nil {
+			return err
+		}
+		// JSON should be of form {"VariantName": {...}}
+		if len(rawMap) != 1 {
+			return errors.New("deserializing ConstraintKind as an enum variant; expecting a single key")
+		}
+		for k, v := range rawMap {
+			variantName = k
+			variantValue = &v
+		}
+	}
+	switch variantName {
+	case "Eq":
+		var variant ConstraintKindEq
+		if variantValue != nil {
+			err := json.Unmarshal(*variantValue, &variant)
+			if err != nil {
+				return err
+			}
+		}
+		*result = ConstraintKind{variant}
+		return nil
+	case "In":
+		var variant ConstraintKindIn
+		if variantValue != nil {
+			err := json.Unmarshal(*variantValue, &variant)
+			if err != nil {
+				return err
+			}
+		}
+		*result = ConstraintKind{variant}
+		return nil
+	case "Contains":
+		var variant ConstraintKindContains
+		if variantValue != nil {
+			err := json.Unmarshal(*variantValue, &variant)
+			if err != nil {
+				return err
+			}
+		}
+		*result = ConstraintKind{variant}
+		return nil
+	case "Neq":
+		var variant ConstraintKindNeq
+		if variantValue != nil {
+			err := json.Unmarshal(*variantValue, &variant)
+			if err != nil {
+				return err
+			}
+		}
+		*result = ConstraintKind{variant}
+		return nil
+	case "Nin":
+		var variant ConstraintKindNin
+		if variantValue != nil {
+			err := json.Unmarshal(*variantValue, &variant)
+			if err != nil {
+				return err
+			}
+		}
+		*result = ConstraintKind{variant}
+		return nil
+	}
+
+	return fmt.Errorf("cannot deserialize ConstraintKind: %s", string(b))
+}
+
+func (variant ConstraintKind) MarshalJSON() ([]byte, error) {
+	switch inner := variant.ConstraintKindVariant.(type) {
+	case ConstraintKindEq:
+		return json.Marshal("Eq")
+	case ConstraintKindIn:
+		return json.Marshal("In")
+	case ConstraintKindContains:
+		return json.Marshal("Contains")
+	case ConstraintKindNeq:
+		return json.Marshal("Neq")
+	case ConstraintKindNin:
+		return json.Marshal("Nin")
+	default:
+		return nil, fmt.Errorf("unexpected variant %#v of %v", inner, variant)
+	}
+
+}
+
+// ConstraintValueTerm newtype
+type ConstraintValueTerm Term
```

Comment:
This is the biggest problem with just automatically converting the core type. The type we want to pass users is almost the same as the core Filter but we want the values to be go values, not a Term object. I think in my manual version I made this an interface{} and had a post parsing pass to call host.toGo() on it. That isn't too great either but we'll have to do something like that.

Reply:
Ahh, okay that makes sense. It might make sense to have _all_ of `polar_types.go` be internal-only, and pass them through `toGo` to get the Go-side type. Most of the `toGo` code will look similar to existing ones: recursively toGo fields/arrays etc. Does allow us to customise things a little if we want.

I'll revert the filter plan changes from this PR though, since it's pretty far from ready based on that.